File: avalon-support/avalon-oauth2-support/avalon-oauth2-common/src/main/java/io/github/lishangbu/avalon/oauth2/common/userdetails/UserInfo.java
package io.github.lishangbu.avalon.oauth2.common.userdetails;

import java.io.Serial;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.SpringSecurityCoreVersion;
import org.springframework.security.core.userdetails.User;
import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal;

/// 用户信息
///
/// 封装 Spring Security User 并实现 OAuth2AuthenticatedPrincipal，支持附加参数
///
/// @author lishangbu
/// @since 2025/8/9
@Getter
@Setter
@SuppressWarnings("removal")
@EqualsAndHashCode(callSuper = true)
public class UserInfo extends User implements OAuth2AuthenticatedPrincipal {
    @Serial
    private static final long serialVersionUID = SpringSecurityCoreVersion.SERIAL_VERSION_UID;

    /// 附加参数：用于在获取 Token 接口返回
    @Getter private final Map<String, Object> additionalParameters = new HashMap<>();

    public UserInfo(
            String username,
            @Nullable String password,
            Collection<? extends GrantedAuthority> authorities) {
        super(username, password, authorities);
    }

    public UserInfo(
            String username,
            @Nullable String password,
            boolean enabled,
            boolean accountNonExpired,
            boolean credentialsNonExpired,
            boolean accountNonLocked,
            Collection<? extends GrantedAuthority> authorities) {
        super(
                username,
                password,
                enabled,
                accountNonExpired,
                credentialsNonExpired,
                accountNonLocked,
                authorities);
    }

    @Override
    public Map<String, Object> getAttributes() {
        return additionalParameters;
    }

    @Override
    @NullMarked
    public String getName() {
        return super.getUsername();
    }

    @Override
    @NullMarked
    public String toString() {
        return getClass().getName()
                + " ["
                + "Username="
                + getUsername()
                + ", "
                + "Enabled="
                + isEnabled()
                + ", "
                + "AdditionalParameters"
                + additionalParameters
                + "AccountNonExpired="
                + isAccountNonExpired()
                + ", "
                + "CredentialsNonExpired="
                + isCredentialsNonExpired()
                + ", "
                + "AccountNonLocked="
                + isAccountNonLocked()
                + ", "
                + "Granted Authorities="
                + getAuthorities()
                + "]";
    }
}

File: avalon-support/avalon-oauth2-support/avalon-oauth2-common/src/main/kotlin/io/github/lishangbu/avalon/oauth2/common/userdetails/UserInfo.kt
package io.github.lishangbu.avalon.oauth2.common.userdetails

import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal

/**
 * OAuth2 用户信息对象
 *
 * 基于 Spring Security 的 [User] 扩展附加属性，便于在令牌相关接口中返回更多用户信息
 *
 * @author lishangbu
 * @since 2025/8/9
 */
@Suppress("removal")
class UserInfo :
    User,
    OAuth2AuthenticatedPrincipal {
    /** 附加属性 */
    val additionalParameters: MutableMap<String, Any> = HashMap()

    constructor(
        username: String,
        password: String?,
        authorities: Collection<out GrantedAuthority>,
    ) : super(username, password, authorities)

    constructor(
        username: String,
        password: String?,
        enabled: Boolean,
        accountNonExpired: Boolean,
        credentialsNonExpired: Boolean,
        accountNonLocked: Boolean,
        authorities: Collection<out GrantedAuthority>,
    ) : super(
        username,
        password,
        enabled,
        accountNonExpired,
        credentialsNonExpired,
        accountNonLocked,
        authorities,
    )

    /** 返回附加属性映射 */
    override fun getAttributes(): MutableMap<String, Any> = additionalParameters

    /** 返回主体名称 */
    override fun getName(): String = username

    /** 返回对象的字符串表示 */
    override fun toString(): String =
        javaClass.name +
            " [" +
            "Username=" +
            username +
            ", " +
            "Enabled=" +
            isEnabled +
            ", " +
            "AdditionalParameters" +
            additionalParameters +
            "AccountNonExpired=" +
            isAccountNonExpired +
            ", " +
            "CredentialsNonExpired=" +
            isCredentialsNonExpired +
            ", " +
            "AccountNonLocked=" +
            isAccountNonLocked +
            ", " +
            "Granted Authorities=" +
            authorities +
            "]"

    companion object {
        /** 序列化版本号 */
        private const val serialVersionUID = 1L
    }
}
